Feat: ceph-admin interface for adopting an external ceph cluster

UtkarshBhatthere opened the pull request:
## Normal Bootstrap Workflow

### Configuration
- No special configuration required
- `wait-to-adopt` config defaults to `False`

### Process Flow

1. **Charm Initialization**
   ```
   __init__() → __post_init__()
   ```
   - Checks `bootstrapped()` → returns False
   - Sets `bootstrap_status` to `MaintenanceStatus("Service not bootstrapped")`

2. **Event Triggers `configure_charm()`**
   ```
   configure_charm(event)
   └── configure_unit(event)
       └── super().configure_unit(event)
           └── Sets _state.unit_bootstrapped = True
   └── configure_app(event)
       └── configure_app_leader(event)  [if leader]
           └── bootstrap_cluster(event)
               └── microceph.bootstrap_cluster()
           └── handle_config_leader_set_ready()
               └── set_leader_ready()
           └── handle_config_leader_ceph_pool_pgs(event)
           └── handle_config_rgw_service(event)
           └── handle_config_leader_charm_upgrade()
           └── handle_config_leader_new_node(event)
   └── bootstrap_status.set(ActiveStatus())
   └── post_config_setup()
   ```

3. **Bootstrap State Flags Set**
   - `_state.unit_bootstrapped = True` (set in `configure_unit()`)
   - `leader_ready` flag set (set in `configure_app_leader()`)
   - `bootstrap_status` set to Active
   - `peers.interface.state.joined = True`

4. **Status Transition**
   ```
   MaintenanceStatus("Service not bootstrapped")
   → ActiveStatus()
   ```
---

## Adopt Bootstrap Workflow

### Configuration
- Requires `wait-to-adopt: True` in charm config
- Requires `adopt-ceph` relation with external Ceph cluster

### Process Flow

1. **Charm Initialization**
   ```
   __init__() → __post_init__()
   ```
   - Checks `bootstrapped()` → returns False
   - Sets `bootstrap_status` to `MaintenanceStatus("Service not bootstrapped")`

2. **Leader Waits for Adoption** (bypasses normal bootstrap)
   ```
   configure_charm(event)
   └── configure_unit(event)
       └── Does NOT set _state.unit_bootstrapped (bypassed)
   └── configure_app(event)
       └── configure_app_leader(event)  [if leader]
           └── if wait-to-adopt:
               └── return early  (no bootstrap)
   ```
   - Leader unit skips bootstrap and waits
   - Status remains: `MaintenanceStatus("Service not bootstrapped")`

3. **Adopt Relation Data Arrives**
   ```
   adopt-ceph relation-changed event
   └── AdoptCephRequires._on_relation_changed()
       └── emit adopt_ceph_bootstrap event
           └── AdoptCephRequiresHandler._on_bootstrap()
               └── Validate relation data (fsid, mon_hosts, admin_key)
               └── charm.adopt_cluster(fsid, mon_hosts, admin_key)
                   └── microceph.adopt_ceph_cluster()
                       └── Sets peers.interface.state.joined = True
               └── callback_f(event)  [handle_ceph_adopt()]
   ```

4. **Post-Adoption Configuration** (handle_ceph_adopt)
   ```
   handle_ceph_adopt(event)
   └── Check is_leader_ready() → False (proceed)
   └── _state.unit_bootstrapped = True  [MANUAL SET]
   └── handle_config_leader_set_ready()
       └── set_leader_ready()
   └── handle_config_leader_ceph_pool_pgs(event)
   └── handle_config_rgw_service(event)
   └── handle_config_leader_charm_upgrade()
   └── handle_config_leader_new_node(event)
   └── bootstrap_status.set(ActiveStatus())  [MANUAL SET]
   └── status.set(ActiveStatus("charm is ready"))  [MANUAL SET]
   ```

5. **Bootstrap State Flags Set**
   - `_state.unit_bootstrapped = True` (set manually in `handle_ceph_adopt()`)
   - `leader_ready` flag set (via `handle_config_leader_set_ready()`)
   - `bootstrap_status` set to Active (manually)
   - `status` (workload status) set to Active (manually)
   - `peers.interface.state.joined = True` (set in `adopt_cluster()`)

6. **Status Transition**
   ```
   MaintenanceStatus("Service not bootstrapped")
   → BlockedStatus("Waiting for fsid, mon_hosts...") [if data incomplete]
   → ActiveStatus("charm is ready")
   ```

UtkarshBhatthere committed:
feat: ceph-admin interface for remote cluster adoption

Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>

UtkarshBhatthere committed:
fix: lint

Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>

UtkarshBhatthere committed:
test: mock CephCOSAgentProvider is_ready_cb for unit tests

Add a centralized mock for the is_ready_cb callback in
CephCOSAgentProvider
to ensure it always returns True during unit tests. This prevents test
failures caused by the charm's ready_for_service() method returning
False
when the leader is not marked as ready in test scenarios.

The mock is applied in the _MicroCephCharm test class constructor by
patching CephCOSAgentProvider.__init__ to intercept and replace the
is_ready_cb parameter. This ensures all unit tests inheriting from the
test base charm benefit from this behavior without requiring individual
test modifications.

Fixes: test_cos_integration
Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>

UtkarshBhatthere committed:
test: configure RGW service when traefik route becomes ready

Ensure RGW configuration is applied when the traefik route becomes
ready, even without a certificate-transfer relation.

The identity-service ready event fires before the id_svc.ready
property becomes True during event processing. By adding an explicit
call to configure_rgw_service() in the traefik handler, we ensure
configuration happens after the identity service has fully initialized.

Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>

UtkarshBhatthere committed:
fix: fix pep8 linting errors

Fix docstring and comment formatting issues:
- Add periods to docstrings (PEP 257)
- Reduce block comment hashes from 5 to 1 (E266)
- Remove extra blank line after docstring (D202)
- Add missing docstring to property method (D102)

Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>

UtkarshBhatthere committed:
refactor: simplify get_relation_handlers using table

Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>

UtkarshBhatthere committed:
test: fix functional test bundle

Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>

UtkarshBhatthere committed:
tests: fix lint and functional tests

Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>

UtkarshBhatthere committed:
fix: mark leader ready after adopt bootstrap

After adopting an external Ceph cluster, the charm remained in
maintenance with "Service not bootstrapped" message instead of
transitioning to active status.

Root cause analysis revealed four issues:
1. handle_ceph_adopt() checked ready_for_service() which returns True
   after adoption, causing it to skip post-bootstrap configuration.
   Fixed by checking is_leader_ready() instead.
2. The unit's _state.unit_bootstrapped flag was never set, which is
   required by bootstrapped() method. The adopt flow bypasses the
   normal configure_unit() call. Fixed by setting it in handle_ceph_adopt().
3. The bootstrap_status was never set to Active. Sunbeam's status pool
   prioritizes maintenance status over active status. Fixed by setting
   bootstrap_status.set(ActiveStatus()) after post-bootstrap config.
4. The workload status remained blocked from when incomplete adopt data
   raised BlockedExceptionError. Fixed by explicitly clearing it with
   status.set(ActiveStatus()) after successful adoption.

Added unit tests to verify the fix using bootstrapped() method.

Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>

UtkarshBhatthere commented on src/charm.py:
```diff
@@ -467,29 +480,19 @@ def configure_app_leader(self, event: ops.framework.EventBase) -> None:
         if not self.is_leader_ready():
+            if self.model.config.get("wait-to-adopt"):
+                logger.info("skipping configure until microceph adopts a remote ceph cluster")
+                return
```
sunbeam has the check for unit/ app bootstrap inbuild, so we do not need to report the status manually.

UtkarshBhatthere committed:
fix: address PR review comments

Addressing review comments from PR #212:

1. Add descriptive docstring to AdoptCephBootstrapEvent explaining when
   it's emitted and what it represents
2. Add defensive check for next(iter(relation.units)) to handle edge case
3. Validate mon_hosts after splitting to ensure non-empty host list
4. Add error handling for int conversion of default-pool-size config
5. Fix grammar in log message: 'charm not ready to process COS events'

Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>

UtkarshBhatthere commented on lib/charms/ceph_mon/v0/ceph_cos_agent.py:
```diff
@@ -50,11 +50,16 @@ def __init__(self, charm, refresh_cb = None, departed_cb = None):
 
+        if self._is_ready_cb and not self._is_ready_cb():
+            # do not proceed if the charm is not ready
+            logger.debug("charm not ready, to process COS events")
```
ack

UtkarshBhatthere committed:
test: fix adopt functional test

Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>

sabaini commented on .github/workflows/build-and-test.yml:
```diff
@@ -943,6 +942,104 @@ jobs:
         if: ${{ failure() && runner.debug }}
         uses: canonical/action-tmate@main
 
+  juju-adopt-ceph-test:
```
Fwiw we have pytest based functests now in tests/functests, nothing to block over but we could consider using these instead of shell

sabaini commented on .github/workflows/build-and-test.yml:
```diff
@@ -943,6 +942,104 @@ jobs:
+      - name: Install dependencies
+        run: ./tests/scripts/ci_helpers.sh install_deps
+
+      - name: Install MicroCeph charm
```
Nit: maybe say "Install MicroCeph and Charmed Ceph" to reflect we're installing both

UtkarshBhatthere replied:
ack

sabaini commented on .github/workflows/build-and-test.yml:
```diff
@@ -943,6 +942,104 @@ jobs:
+        run: |
+          set -eux
+          date
+          juju integrate ceph-mon primary
```
Nit: to make it super clear which interface we're adding maybe specify the full rel name

sabaini commented on lib/charms/ceph_mon/v0/ceph_cos_agent.py:
```diff
@@ -50,11 +50,16 @@ def __init__(self, charm, refresh_cb = None, departed_cb = None):
+        if self._is_ready_cb and not self._is_ready_cb():
+            # do not proceed if the charm is not ready
+            logger.debug("charm not ready to process COS events")
+            return
```
Do we need to defer() the event here?

sabaini commented on src/microceph_adopt_ceph.py:
```diff
@@ -0,0 +1,177 @@
+                fsid = remote_ceph_data.get(AdoptCephRelationDataKeys.fsid.value, None)
+                admin_key = remote_ceph_data.get(AdoptCephRelationDataKeys.admin_key.value, None)
+
+                logger.debug(
```
Nit: important lifecycle info, would make this logger.info() (no pun intended)

UtkarshBhatthere replied:
ack

sabaini reviewed (changes requested):
Hey @UtkarshBhatthere thanks for this chunk of work :blush: 
Some questiosn and nitpickery inline

UtkarshBhatthere commented on src/microceph_adopt_ceph.py:
```diff
@@ -0,0 +1,177 @@
+
+        self.framework.observe(self.adopt_ceph.on.adopt_ceph_bootstrap, self._on_bootstrap)
+
+    def _on_bootstrap(self, relation):
```
changed them to `relation_event`

UtkarshBhatthere commented on .github/workflows/build-and-test.yml:
```diff
@@ -943,6 +942,104 @@ jobs:
+          juju wait-for application secondary --query='forEach(units, unit => unit.workload-status=="active" && unit.agent-status=="idle")' --timeout=30m
+          date
+
+      - name: Adopt the charmed ceph cluster using primary microcpeh application
```
ack

UtkarshBhatthere committed:
fix: Resolve review comments

Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>

UtkarshBhatthere commented on .github/workflows/build-and-test.yml:
```diff
@@ -943,6 +942,104 @@ jobs:
+          set -eux
+          date
+          juju integrate ceph-mon primary
+          juju wait-for unit primary/0 --query='workload-status=="active"' --timeout=15m
```
No, when deployed with "wait-for-adopt", the charm does not go to active state rather, it goes to `Waiting` with `Service not bootstrapped`. I will add that as a wait for expectation as well.

UtkarshBhatthere committed:
fix: make the adopt bootstrap handler sturdy

Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>

UtkarshBhatthere commented on src/microceph_adopt_ceph.py:
```diff
@@ -0,0 +1,177 @@
+        self.on.adopt_ceph_bootstrap.emit(event.relation)
+
+    def _on_relation_broken(self, event) -> None:
+        """On relation departed."""
```
rewrote the method descriptions for broken and changed event handlers.

UtkarshBhatthere commented on src/microceph_adopt_ceph.py:
```diff
@@ -0,0 +1,177 @@
+                    "Adopt relation removed before cluster bootstrap could be performed"
+                )
+
+        logger.debug("Ignoring adopt-ceph departed event")
```
ack

UtkarshBhatthere committed:
test: add missing mocks for test_handle_ceph_adopt_marks_leader_ready

Added mocks for ceph.enable_mgr_module and ceph_cos_agent.ceph_utils
to prevent command execution during tests.

Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>

UtkarshBhatthere committed:
fix: resolve lint issues

Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>

UtkarshBhatthere committed:
test: fix endpoint for adopt integration

Signed-off-by: Utkarsh Bhatt <utkarsh.bhatt@canonical.com>

sabaini commented:
Just as a quick update, liking the changes so far, only thing we could maybe add is hiding keys cf. https://github.com/canonical/charm-microceph/pull/212#discussion_r2590146903 :+1:

sabaini reviewed (approved):
Two failing functests are known pre-existing CI issues -- change lgtm!
Thanks @UtkarshBhatthere massive piece of work
